### PR TITLE
Added support to have one catalog name pointing to multiple data store shards.

### PR DESCRIPTION
**Pull request by ajoymajumdar**
Example: We can add multiple catalogs, referencing multiple datasources, addressed by the same catalog name.

### CODE REVIEW

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/spi/MetacatCatalogConfig.java [hunk 1]
```diff
@@ -104,6 +89,17 @@ public boolean isThriftInterfaceRequested() {
...
     public static class Keys {
+        /**
+         * Catalog name. Usually catalog name is derived from the name of the catalog properties file.
+         * One could also specify it in the properties under the property name <code>catalog.name</code>.
+         * For example, if a catalog has two shards with catalog defined in two property files,
+         * then we can have the <code>catalog.name</code> set to one name.
+         */
+        public static final String CATALOG_NAME = "catalog.name";
+        /**
+         * Connector type.
```
Comment (zhljen) [inline metacat-common-server/src/main/java/com/netflix/metacat/common/server/spi/MetacatCatalogConfig.java]: is this type or name?

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorContext.java [hunk 2]
```diff
@@ -30,6 +30,14 @@
...
      */
     private final String catalogName;
+    /**
+     * Catalog shard name.
+     */
+    private final String catalogShardName;
+    /**
+     * Catalog type.
+     */
+    private final String type;
```
Comment (zhljen) [inline metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorContext.java]: We should name the type explicitly as connectorType.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/util/DataSourceManager.java [hunk 3]
```diff
@@ -50,13 +50,13 @@ public static DataSourceManager get() {
...
-     * @param catalogName catalog name
+     * @param name catalog name
      * @param properties  properties
      * @return DataSourceManager
      */
-    public DataSourceManager load(final String catalogName, final Map<String, String> properties) {
-        if (dataSources.get(catalogName) == null) {
-            createDataSource(catalogName, properties);
+    public DataSourceManager load(final String name, final Map<String, String> properties) {
+        if (dataSources.get(name) == null) {
```
Comment (zhljen) [inline metacat-common-server/src/main/java/com/netflix/metacat/common/server/util/DataSourceManager.java]: what is this name? If this is not 'catalogName' anymore, we should change the comments 
`* @param name catalog name` to the right one.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/util/DataSourceManager.java [hunk 4]
```diff
@@ -78,15 +78,15 @@ public DataSourceManager load(final String catalogName, final Properties propert
     /**
      * Returns the data source loaded for the given catalog.
      *
-     * @param catalogName catalog name
+     * @param name catalog name
      * @return DataSource
```
Comment (zhljen) [inline metacat-common-server/src/main/java/com/netflix/metacat/common/server/util/DataSourceManager.java]: same as above.

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/CatalogManager.java [hunk 5]
```diff
@@ -38,7 +38,15 @@
...
 
 /**
- * Catalog manager.
+ * Catalog manager. This loads the catalogs defined as .properties files under the location defined by config property
+ * <code>metacat.plugin.config.location</code>.
+ * Usually there is a one-to-one mapping between a catalog and a data store. We could also have data stores
+ * (mostly sharded) addressed by a single catalog name.
+ * If a data store is sharded and needs to be represented by a single catalog name, then we will have multiple catalog
+ * property files, each referencing to its physical data store and having the same <code>catalog.name</code>.
+ * In this case, <code>catalog.name</code> and <code>metacat.schema.whitelist</code> will be used to point to the right
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/manager/CatalogManager.java]: schema. whitelist?  Why not using 'shards'?

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/CatalogManager.java [hunk 6]
```diff
@@ -114,16 +124,22 @@ private void loadCatalog(final File file) throws Exception {
...
+        // If multiple catalog property files refer the same catalog name, then there will be multiple shard names
+        // with the same catalog name.
+        final String catalogShardName = Files.getNameWithoutExtension(file.getName());
+        // If catalog name is not specified, then use the catalog shard name.
+        final String catalogName = properties.getOrDefault(MetacatCatalogConfig.Keys.CATALOG_NAME, catalogShardName);
+        final ConnectorContext connectorContext =
+            new ConnectorContext(catalogName, catalogShardName, connectorType, config, registry, properties);
+        this.connectorManager.createConnection(connectorContext);
         log.info("-- Added catalog {} using connector {} --", catalogName, connectorType);
     }
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/manager/CatalogManager.java]: are we allowing one file having multiple shards configs? Also, can we log the shard explicitly? 
`log.info("-- Added catalog {} using connector {} --", catalogName, connectorType);`

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java [hunk 7]
```diff
@@ -103,31 +116,91 @@ public void addPlugin(final ConnectorPlugin connectorPlugin) {
...
-
             final MetacatCatalogConfig catalogConfig =
-                MetacatCatalogConfig.createFromMapAndRemoveProperties(connectorType,
+                MetacatCatalogConfig.createFromMapAndRemoveProperties(connectorType, catalogName,
                     connectorContext.getConfiguration());
-            catalogs.put(catalogName, catalogConfig);
+            final List<String> databaseNames = catalogConfig.getSchemaWhitelist();
+            if (databaseNames.isEmpty()) {
+                Preconditions.checkState(!catalogs.contains(catalogName, EMPTY_STRING),
+                    "A catalog with name %s already exists", catalogName);
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java]: What if the catalog doesn't exist? Should we logging error for this case?

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java [hunk 8]
```diff
@@ -103,31 +116,91 @@ public void addPlugin(final ConnectorPlugin connectorPlugin) {
...
+                log.debug("Catalog {} doesn't support getTableService. Ignoring.", catalogName);
+            }
+            try {
+                partitionServices.add(connectorFactory.getPartitionService());
+            } catch (UnsupportedOperationException e) {
+                log.debug("Catalog {} doesn't support getPartitionService. Ignoring.", catalogName);
+            }
+            final CatalogHolder catalogHolder = new CatalogHolder(catalogConfig, connectorFactory);
+            if (databaseNames.isEmpty()) {
+                catalogs.put(catalogName, EMPTY_STRING, catalogHolder);
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java]: Maybe this should be moved to the top along with the first
`if (databaseNames.isEmpty()) {`

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java [hunk 9]
```diff
@@ -41,26 +42,39 @@
...
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
 
 /**
  * Connector manager.
  */
 @Slf4j
 public class ConnectorManager {
+    private static final String EMPTY_STRING = "";
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java]: Actually. this (ConnectorManager) looks more like a catalog manager and the catalogManger is more like a connectorManager. We should switch their name.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorFactory.java [hunk 10]
```diff
@@ -57,12 +57,19 @@ default ConnectorPartitionService getPartitionService() {
     }
 
     /**
-     * Returns the name of the connector.
+     * Returns the name of the catalog.
      *
-     * @return Returns the name of the connector.
+     * @return Returns the name of the catalog.
      */
     String getName();
```
[the remark this answers is not in the document]
Comment (ajoymajumdar) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorFactory.java]: Consider Connector as the definition and Catalog has an instance of it. Multiple catalogs can have one connector. I have kept the method name as it is but modified the comment to clear the confusion that was raised in our discussion.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorContext.java (hunk 2 above)
Comment (ajoymajumdar) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorContext.java]: Will do.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/spi/MetacatCatalogConfig.java (hunk 1 above)
Comment (ajoymajumdar) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/spi/MetacatCatalogConfig.java]: This is the connector type. We had this property as connector.name. Probably we had this inherited from presto.

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/util/DataSourceManager.java (hunk 3 above)
Comment (ajoymajumdar) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/util/DataSourceManager.java]: Can be anything, this is a generic class. Consumers can use any name - catalog name or catalog shard name.

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/CatalogManager.java (hunk 5 above)
Comment (ajoymajumdar) [reply metacat-main/src/main/java/com/netflix/metacat/main/manager/CatalogManager.java]: This was already defined. Concept of shards is higher than this. A shard will contain a list of databases.

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java (hunk 9 above)
Comment (ajoymajumdar) [reply metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java]: CatalogManager loads catalogs. ConnectorManager provides the right connector properties/services.

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java (hunk 8 above)
Comment (zhljen) [reply metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java]: this is empty represents everything ( as default) or a real empty database?  Are we use empty string as a placeholder for the catalogs having no database?

File: metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java [hunk 11]
```diff
@@ -238,4 +328,14 @@ public ConnectorTypeConverter getTypeConverter(final String connectorType) {
     public ConnectorInfoConverter getInfoConverter(final String connectorType) {
         return getPlugin(connectorType).getInfoConverter();
     }
+
+    /**
+     * A Holder class holding the catalog's config and connector factory.
+     */
+    @Data
+    @AllArgsConstructor
+    private static class CatalogHolder {
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java]: I feel we should factor out the catalogRelated operation into a new class, separating from the database/table/part services related informations. The catalog class has 
`private final Table<String, String, CatalogHolder> catalogs = HashBasedTable.create();`
and catalog related information.

Comment (zhljen) [reply metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java]: The catalogHolder is not needed if the table has four columns: catalogName, databaseName, config, connectorFactory

File: metacat-main/src/main/java/com/netflix/metacat/main/services/impl/CatalogServiceImpl.java [hunk 12]
```diff
@@ -70,20 +71,29 @@ public CatalogServiceImpl(
...
-                .collect(Collectors.toList())
-        );
+        final List<String> databases = Lists.newArrayList();
+        configs.forEach(config -> {
+            QualifiedName qName = name;
+            if (config.getSchemaWhitelist().isEmpty()) {
+                result.setType(config.getType());
+            } else {
+                qName = QualifiedName.ofDatabase(name.getCatalogName(), config.getSchemaWhitelist().get(0));
+            }
```
Comment (zhljen) [inline metacat-main/src/main/java/com/netflix/metacat/main/services/impl/CatalogServiceImpl.java]: Don't understand here. If always get the first one, why do we need list?

File: metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorFactory.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (zhljen) [reply metacat-common-server/src/main/java/com/netflix/metacat/common/server/connectors/ConnectorFactory.java]: I fee Connector is a mechanism to load catalog, which is more implementation details. Catalog is the concept/definition existing in the metacat architecture.

**Pull request closed without merging**